game-core/src/main/java/games/strategy/triplea/ui/screen/TileManager.java

```diff
@@ -371,6 +377,23 @@ private void drawUnits(final Territory territory, final MapData mapData, final S
     }
   }
 
+  private static List<UnitCategory> getSortedUnitCategories(final Territory t, final GameData data) {
+    final List<UnitCategory> categories = new ArrayList<>(UnitSeperator.categorize(t.getUnits().getUnits()));
+    final List<PlayerID> players = data.getPlayerList().getPlayers();
+    if (players.contains(t.getOwner())) {
+      players.remove(t.getOwner());
+      players.add(0, t.getOwner());
+    }
+    final List<UnitType> xmlUnitTypes = new ArrayList<>(data.getUnitTypeList().getAllUnitTypes());
+    categories.sort(Comparator
+        .comparing(UnitCategory::getOwner, Comparator.comparing(players::indexOf))
+        .thenComparing(UnitCategory::getType, Comparator.comparing(ut -> !UnitAttachment.get(ut).getIsInfrastructure()))
+        .thenComparing(UnitCategory::getType, Comparator.comparing(ut -> !Matches.unitTypeIsLand().test(ut)))
+        .thenComparing(UnitCategory::getType, Comparator.comparing(ut -> !UnitAttachment.get(ut).getIsSea()))
+        .thenComparing(UnitCategory::getType, Comparator.comparing(xmlUnitTypes::indexOf)));
```

Comment:
The amount of overall code can be reduced by nesting specific calls like this:
```java
categories.sort(Comparator
  .comparing(UnitCategory::getOwner, Comparator.comparing(players::indexOf))
  .thenComparing(UnitCategory::getType, Comparator
    .comparing(ut -> !UnitAttachment.get(ut).getIsInfrastructure())
    .thenComparing(ut -> !Matches.unitTypeIsLand().test(ut))
    .thenComparing(ut -> !UnitAttachment.get(ut).getIsSea())
    .thenComparing(xmlUnitTypes::indexOf));
```

Reply:
Ended up adding some additional logic as well as cleaning this up a bit. Only thing I don't like is that Java can't seem to handle the type inference in the chaining so need to explicitly declare the type of the key extractor. If there is some better way please let me know.